source/features/wait-for-build.tsx

```diff
@@ -55,62 +69,122 @@ function disableForm(disabled = true): void {
 }
 
 async function handleMergeConfirmation(event: delegate.Event<Event, HTMLButtonElement>): Promise<void> {
-	const checkbox = getCheckbox();
-	if (checkbox?.checked) {
-		event.preventDefault();
+	if (!getCheckbox()?.checked) {
+		return;
+	}
+
+	const lastCommitSha = prCiStatus.getLastCommitReference()?.trim();
+	if (!lastCommitSha) {
+		return;
+	}
 
-		disableForm();
-		const currentConfirmation = Symbol('');
-		waiting = currentConfirmation;
-		const status = await prCiStatus.wait();
+	event.preventDefault();
+	disableForm();
+	const currentConfirmation = Symbol('');
+	waiting = currentConfirmation;
 
-		// Ensure that it wasn't cancelled/changed in the meanwhile
-		if (waiting === currentConfirmation) {
-			disableForm(false);
+	let result: prCiStatus.CommitStatus;
+	try {
+		result = await pRetry(async () => {
+			const status = await prCiStatus.getCommitStatus(lastCommitSha);
 
-			if (status === prCiStatus.SUCCESS) {
-				event.delegateTarget.click();
+			// Ensure that it wasn't cancelled/changed in the meanwhile
+			if (waiting !== currentConfirmation) {
+				throw new AbortError('The merge was cancelled or a new commit was pushed');
 			}
-		}
+
+			if (status === prCiStatus.PENDING) {
+				throw new Error('CI is not done yet');
+			}
+
+			return status;
+		}, {
+			forever: true,
+			minTimeout: 5000,
+			maxTimeout: 10_000,
+		});
+	} catch {
+		return;
+	} finally {
+		disableForm(false);
+	}
+
+	if (result === prCiStatus.SUCCESS) {
+		event.delegateTarget.classList.add('rgh-merging'); // Avoid triggering the event listener again
+		event.delegateTarget.click();
 	}
 }
 
+function watchForNewCommits(): VoidFunction {
+	let previousCommit = prCiStatus.getLastCommitReference();
+	const filteredListener = (): void => {
+		const newCommit = prCiStatus.getLastCommitReference();
+		if (newCommit === previousCommit) {
+			return;
+		}
+
+		previousCommit = newCommit;
+		// Cancel submission if a new commit was pushed
+		disableForm(false);
+		showCheckboxIfNecessary();
+	};
+
+	const observer = observeElement('.js-discussion', filteredListener, {
+		childList: true,
+		subtree: true,
+	})!;
+
+	return observer.disconnect;
+}
+
 function onBeforeunload(event: BeforeUnloadEvent): void {
 	if (waiting) {
 		event.returnValue = '';
 	}
 }
 
-function init(): VoidFunction {
-	// Watch for new commits and their statuses
-	prCiStatus.addEventListener(showCheckboxIfNecessary);
+async function init(): Promise<false | VoidFunction[]> {
+	if (await getWorkflowsCount() === 0) {
+		return false;
+	}
 
 	onPrMergePanelOpen(showCheckboxIfNecessary);
 
-	// One of the merge buttons has been clicked
-	delegate(document, '.js-merge-commit-button', 'click', handleMergeConfirmation);
+	const deinit: VoidFunction[] = [
+		watchForNewCommits(),
```

Comment:
This adds an observer on the page of every open PR, even if the user is just here to read the conversation and has no intention of merging yet.

This should probably be added only when the user first opens the merge box.

Reply:
We have a listener for the merge box opening, we should use it